### DISCUSSION TITLE
Is there any ways to get an Array from the json file directly?

### DISCUSSION THREAD

**Discussion by DongwonTTuna** [robisim74/qwik-speak#25] [General]
Hello! 😊

I appreciate you for creating this great library for qwik. It has been very helpful to me.

I am aware that there is a method to parse strings into arrays, but I would like to inquire if there is a way to retrieve an array from a JSON file.

the JSON file is.
```
{
  "common": {
    "slides": ["visual4.png","visual1.png","visual2.png","visual3.png"]
  }
}
```

In the next-i18next library, I have been able to achieve this using 
`Array.from(t("slides", { returnObjects: true }))`

Also, I would like to know if there are any possible ways to save an array as a default value directly to JSON?

Currently, I am saving it in the following way:
```
{t("common.slides@@visual4.png,visual1.png,visual2.png,visual3.png")
.split(",")
.map((item, index) => {return (<div>{blablabla}</div>)})

```

Thank you for reading!

**Comment by robisim74** [comment]
Hi @DongwonTTuna ,

Currently, only arrays of keys are supported out of the box. But I think it would be possible to implement `key => array` as well according to the syntax you suggested:

**from (jsx)**

`const myArray = t("common.slides@@visual1.png,visual2.png");`

**to (json)**
```
{
  "common": {
    "slides": ["visual1.png","visual2.png"]
  }
}
```

**rendered as**

`const myArray = ["visual1.png","visual2.png"];`

**inlined as**

`const myArray = $lang('it-IT') && ["visual1-it.png","visual2-it.png"] || ["visual1.png","visual2.png"];`

If you like it, I'll try to do some testing and add `key => array` support in the next version of the library.

**Reply by DongwonTTuna** [reply]
Thank you for taking the time to respond despite your busy schedule.

I'm sorry, but the result I got after writing and running the code you provided was different from what I expected.
First, I entered the following code you told me:
`const myArray = t("common.slides@@visual1.png,visual2.png");`

like this picture.
<img width="456" alt="スクリーンショット 2023-02-27 21 34 07" src="https://user-images.githubusercontent.com/38794795/221567740-87b5be02-77b9-4e5f-b900-ee581b456b28.png">

and ran it exactly as is.
The following image shows the result of the execution.

<img width="251" alt="スクリーンショット 2023-02-27 21 34 11" src="https://user-images.githubusercontent.com/38794795/221567774-095a1a01-c6a8-4905-9651-e12a5bb3822c.png">

As you can see, instead of an array, the default value after @@ was returned.
I changed it to a string as shown in the next image and ran it again.

<img width="342" alt="スクリーンショット 2023-02-27 21 45 47" src="https://user-images.githubusercontent.com/38794795/221568038-063ddba5-86fb-4ce3-9192-fd8859916b46.png">
<img width="260" alt="スクリーンショット 2023-02-27 21 46 53" src="https://user-images.githubusercontent.com/38794795/221568060-75c0e096-462b-4557-abda-e43ff553e32f.png">

This time, the output was as expected.
I also confirmed that when running `qwik-speak-extract`, the values in the array were overwritten with the default value likes below image.

<img width="279" alt="スクリーンショット 2023-02-27 21 49 44" src="https://user-images.githubusercontent.com/38794795/221568192-a4fb1f3f-498b-47b5-953d-7ac6121dfc88.png">



I'm sorry to bother you while you're busy, but I'm wondering if there's something wrong with my code.

Thank you for reading this far.

**Reply by robisim74** [reply]
No problem, but the code I provided is just a proposal based on your syntax.

As I said above, I may implement it in the next version of the library.

Rather, there is the problem of using the _comma_ as an array value separator: it is already a valid character in the default value. Maybe we could directly supply the array as default value, something like this:

`const myArray = t("common.slides@@['visual1.png','visual2.png']");`

What do you think?

**Reply by DongwonTTuna** [reply]
Thank you for your response.

Firstly, I apologize for my misunderstanding. I understand now that it is not currently a available feature.

I have also thought of various solutions, and I'm curious to know your thoughts on using **\` ${}\`** and the **JSON.stringify()** function to convert Array or Object to strings and pass them as default values.

The following pictures show the declaration of each variable and its result:

<img width="501" alt="スクリーンショット 2023-02-28 4 32 15" src="https://user-images.githubusercontent.com/38794795/221664316-3d2e6991-3b1c-4695-a658-62b6c5addd80.png">
<img width="330" alt="スクリーンショット 2023-02-28 4 23 24" src="https://user-images.githubusercontent.com/38794795/221662453-c3bbf35c-16c0-4eb7-98ce-0e32761f6825.png">

This approach is very intuitive and has the advantage of being no different from the current usage. 
Additionally, when loaded, they can be used as an Array or Object by using **JSON.parse()**.

The following pictures explain this.

<img width="458" alt="スクリーンショット 2023-02-28 4 18 35" src="https://user-images.githubusercontent.com/38794795/221661978-c2f43177-cb17-4617-b09e-838c1d2c831e.png">
<img width="554" alt="スクリーンショット 2023-02-28 4 16 48" src="https://user-images.githubusercontent.com/38794795/221662024-f3c70cbb-4a5e-41a2-a0b3-15cf577ae5e4.png">

**Reply by robisim74** [reply]
Your solution is interesting, and it might work fine at runtime, but it wouldn't work with the extraction tool.

_The extraction tool skips by design all strings containing variables_: the tool does a static analysis of the `$translate` function, and would not be able to recover the value of the variables (which could be dynamic, i.e. generated at runtime).

However, I could modify the tool to do the `JSON.parse` behind the scenes, but the default value _must contain the stringified array or object:_

**jsx**
```jsx
const tArray = t('array@@["1", "2", "3", "4"]');
const tObject = t('obj@@{"one": "a", "two": "b"}');
```

**extracted json**
```json
  {
    "array": [
      "1",
      "2",
      "3",
      "4"
    ],
    "obj": {
      "one": "a",
      "two": "b"
    }
  }
```

**rendered as array/object**
```jsx
const tArray = ["1", "2", "3", "4"];
const tObject = {one: "a", two: "b"};
```

Taking your initial example, you would be able to do in your template:
```jsx
    {t('common.slides@@["visual4.png","visual1.png","visual2.png","visual3.png"]')
      .map((item, index) => {return (<div>{blablabla}</div>)})}
```

**Reply by DongwonTTuna** [reply]
Thank you for your response.

I understand. I need to write the Array as a string so that the extractor can recognize it and extract it as a Json file.

I think what you wrote is perfect for any cases.

Thank you for continuously updating the library.😊

**Reply by robisim74** [reply]
@DongwonTTuna In v0.7.0 I added support for arrays/objects. See here: https://github.com/robisim74/qwik-speak/blob/main/docs/translate.md

**Reply by DongwonTTuna** [reply]
Thank you for fast update 🤩
I'll try it today.

**Reply by DongwonTTuna** [reply]
Hello, thank you for updating.😊

I have tested it just now and in most cases, it seems to be working properly.

However, there is one case where it did not work, specifically with 
```typescript
const ArrayObjects = t<Translation[]>('app.ArrayObjects@@[{"one":"1","two":"2"},{"one":"1","two":"2"},{"one":"1","two":"2"}]').
```

and the output is here.

```json
"ArrayObjects": "[{\"one\":\"1\",\"two\":\"2\"},{\"one\":\"1\",\"two\":\"2\"},{\"one\":\"1\",\"two\":\"2\"}]",
```

And it is saved as a string in the JSON format at qwik-speak-extract.
Serious problem is It could not be built.

**Reply by DongwonTTuna** [reply]
I have made it work using `<Translation>` instead of` <Translation[]>`.

Below is an example I have created:

```typescript
const ArrayObjects = t<Translation>('app.ArrayObjects@@{"array":[{"one":"1","two":"2"},{"one":"1","two":"2"},{"one":"1","two":"2"}]}')
```


app.json
```json
{
   "ArrayObjects":{
      "array":[
         {
            "one":"1",
            "two":"2"
         },
         {
            "one":"1",
            "two":"2"
         },
         {
            "one":"1",
            "two":"2"
         }
      ]
   }
}
```

but It could not be built.
How can I fix it?

<img width="463" alt="スクリーンショット 2023-03-06 13 22 34" src="https://user-images.githubusercontent.com/38794795/223018985-c6b8ee22-9419-4e5d-85d4-3031c1cd287f.png">

**Reply by DongwonTTuna** [reply]
ok, I found the bug is caused by updating libary to 0.7.0

I don't know why it is but when I downgrade to 0.6.0, It works.

**Reply by robisim74** [reply]
Thanks for your feedback! There was a bug with array of objects (downgrade to v.0.6.0 it's not an option, because objects were not supported yet).

I released the patch v.0.7.1 which should fix the bug.

Now you should be able to use:
```
const ArrayObjects = t<Translation[]>('app.ArrayObjects@@[{"one":"1","two":"2"},{"one":"1","two":"2"},{"one":"1","two":"2"}]')
```

**Reply by DongwonTTuna** [reply]
Thank you for quickly updating again!🤩

I tried to use it immediately, but it still causes errors.

The problems are as follows:

- The `npm run build` does not work. (same error as the previous version)
- The `qwik-speak-extract` command applies the value of a different file instead of the default value. (See the picture below for details)

The following picture shows the contents of the json file that I used as the default value and that is set as the Default Language.
I set **Korean** as the default language because **English** is **not** my native language.

```typescript
  const imageArray = t<Translation[]>('product.hycera@@["v1677989410/bioden/hycera1_gjjzdt","v1677989411/bioden/hycera2_v46x7l","v1677989411/bioden/hycera3_lqhmo5"]');
```
i18n/ko/product.json
```json
{
  "product":{
    "hycera": [
      "v1677989410/bioden/hycera1_eng_up5keo",
      "v1677989411/bioden/hycera2_eng_hu8rcp",
      "v1677989411/bioden/hycera3_eng_fmuw4c"
    ]
  }
}
```

However, as you can see, the string in ko/product.json was passed as the default value for the hycera array in en/product.json, not the value behind @@.

The value should have been as follows.

i18n/ko/product.json
```json
{
  "product":{
    "hycera": [
      "v1677989410/bioden/hycera1_gjjzdt",
      "v1677989411/bioden/hycera2_v46x7l",
      "v1677989411/bioden/hycera3_lqhmo5"
    ]
  }
}
```


I am not sure of the exact reason why this error occurred, but there were also cases where the array value in cn/product.json became the default value.

In my opinion, during the build process, it is decoded and static pages are created for each language, but with the addition of this feature, there seems to be an error in that process.

Thank you for your time. I appreciate your help 😊

**Reply by robisim74** [reply]
**About the extraction** If I understood correctly, do you expect the default value to override the value in the json file? The default value is used in dev mode and during extraction for any language _where the value doesn't exist in the json file._ If it already exists, of course it is not overwritten, otherwise you would lose the translations already made. If you intend to regenerate it into the json file, you should delete it (just the value or the whole file) before running the `qwik-speak-extract `command.

**About the build** I did a repro: https://stackblitz.com/edit/qwik-speak-vcw7wz?file=src%2Froutes%2F[...lang]%2Findex.tsx where an array is used (already extracted in json files, because extract command doesn't work directly in StackBlitz). If you stop the server, and do `npm run preview`, you'll see that it works. Could you try to recreate the problem, and open a issue?

Thanks!

**Reply by DongwonTTuna** [reply]
Hello, thank you for your response.😊

To reproduce this error, I have created a simple qwik project at the following GitHub repository: https://github.com/DongwonTTuna/qwik-speak-ErrExample.git.

1. Firstly, please check the json files for each language. 
2. Next, please run `npm install && npm run dev` to confirm that it works as expected. 
3. Then, please try running `npm run build` to confirm that an error occurs. 
4. Finally, please use `npm run qwik-speak-extract` to confirm that the contents of each json file are overwritten with values in a different language other than the default values. (**this may need to run the command several times with original i18n files**)

Thank you!
